### ISSUE TITLE
App crashes with exception: java.lang.IllegalStateException: Reply already submitted

### ISSUE THREAD

**Issue by gopchi** [transistorsoft/flutter_background_geolocation#130]
## Your Environment
* Plugin version: 1.2.4
* Platform: Android
* OS version: 7.1.1 / 9
* Device manufacturer / model: Galaxy J5(2016) / Galaxy S9
* Flutter info (`flutter doctor`):

[✓] Flutter (Channel stable, v1.9.1+hotfix.4, on Mac OS X 10.14.6 18G95, locale
    en-US)
[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
[✓] Xcode - develop for iOS and macOS (Xcode 10.2)
[✓] Android Studio (version 3.5)
[✓] VS Code (version 1.39.1)

* Plugin config:
```dart
    state = await bg.BackgroundGeolocation.ready(bg.Config(
      desiredAccuracy: bg.Config.DESIRED_ACCURACY_HIGH,
      disableElasticity: true,
      distanceFilter: 1.0,
      startOnBoot: true,
      logLevel: bg.Config.LOG_LEVEL_OFF,
      activityRecognitionInterval: 0,
      preventSuspend: true,
      reset: true,
      forceReloadOnBoot: true,
      forceReloadOnLocationChange: true,
      maxRecordsToPersist: 5,
      stopOnTerminate: false,
      minimumActivityRecognitionConfidence: 42,
      notificationTitle: 'MyApp',
      notificationText: 'Location services active',
    ));
```

## Expected Behavior

Expected behavior is the app not crashing randomly.

## Actual Behavior

At some point in time the app crashes.

## Steps to Reproduce

1.  Open app
2. Wait a couple of seconds/minutes with even no activity and then it crashes.
3.
4.

## Context

Nothing, just run the app and then wait. This does not happen on all devices. The Samsung Galaxy J5 (2016) used an older OS version and the app worked fine, but immediately after the new OS update came and installed it the app started crashing randomly. On the Samsung S9 it was always like that. 

**The logs are from Firebase Crashlitycs**

## Debug logs

<details>
	<summary>Logs</summary>

``` <!-- syntax-highligting:  DO NOT REMOVE -->
Fatal Exception: java.lang.IllegalStateException: Reply already submitted
       at io.flutter.embedding.engine.dart.DartMessenger$Reply.reply(DartMessenger.java:151)
       at io.flutter.plugin.common.MethodChannel$IncomingMethodCallHandler$1.success(MethodChannel.java:225)
       at com.transistorsoft.flutter.backgroundgeolocation.FLTBackgroundGeolocationPlugin$7.onLocation(FLTBackgroundGeolocationPlugin.java:390)
       at com.transistorsoft.locationmanager.location.TSLocationManager$c$1.run()
       at android.os.Handler.handleCallback(Handler.java:751)
       at android.os.Handler.dispatchMessage(Handler.java:95)
       at android.os.Looper.loop(Looper.java:154)
       at android.app.ActivityThread.main(ActivityThread.java:6816)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1563)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1451)
```

</details>

**Comment by christocracy** [comment]
You cannot be using `1.2.4` as stated since [FLTBackgroundGeolocationPlugin$7.onLocation(FLTBackgroundGeolocationPlugin.java:390)](https://github.com/transistorsoft/flutter_background_geolocation/blob/master/android/src/main/java/com/transistorsoft/flutter/backgroundgeolocation/FLTBackgroundGeolocationPlugin.java#L390) is an empty line.

**Comment by christocracy** [comment]
Sorry, that was master I was looking at.  This is [FLTBackgroundGeolocationPlugin.java:390](https://github.com/transistorsoft/flutter_background_geolocation/blob/1.2.4/android/src/main/java/com/transistorsoft/flutter/backgroundgeolocation/FLTBackgroundGeolocationPlugin.java#L390)

**Comment by christocracy** [comment]
How are you using `#getCurrentPosition`?

**Comment by fpfilipov94** [comment]
> How are you using `#getCurrentPosition`?

I am also a member of the team working on this app, so I will clarify this in more detail.

As you can see, we have configured the plugin to reboot the Dart portion of the app on location changes. This is because that logic needs to work in the background the entire time, as long as the GPS sensor is enabled. Also, because we have very specific custom logic to work with that location data locally after we obtain it, we can't use some of the plugin's other features like the one that allows us to set an url for sending that location data.

So, to meet our requirements, we call `#getCurrentPosition` every 1s using `Timer.periodic`. This is necessary because the app simply needs a stream of fresh location data, even if the device is not moving and the reported location has not changed for a while.

**Comment by christocracy** [comment]
>So, to meet our requirements, we call #getCurrentPosition every 1s using Timer.periodic

That sounds insane.  Do you own these devices?  You're going to kill the user's battery.

Show me your code executing `#getCurrentPosition`.

**Comment by gopchi** [comment]
I know it sounds insane but the app that was ordered to us was also insane. It has a requirement to track user’s location constantly and push it to a real-time database. 

Regarding the devices - yes we own them and the app is still in testing phase, so nobody has been hurt.

Back to the main topic - could you please suggest why is the crash occurring? Why does the 1 second timer even matter when the method should return the current position? Calling such a method that often should not behave like that in my opinion. 

For example in Android you have the last known position and you can always return that as a value no matter how often you call it.


My colleague will send the code tomorrow and you would be able to see it but I really don’t think that the timer is the problem.

**Comment by christocracy** [comment]
>  It has a requirement to track user’s location constantly and push it to a real-time database.

I suggest you argue the requirements.  You can’t develop mobile apps to operate like a web app running on a desktop plugged into ac lower.  Your app is useless after it runs out of power.

Are those who designed the specs aware that this plugin *knows* when the device is moving vs stationary, that it automatically engages tracking when the device is moving, regardless of terminated or device rebooted.

Forget about #getCurrentPosition on a timer.  You might as well just #ready the plugin with ```
disableStopDetection: true, 
distanceFilter: 0, 
locationUpdateInterval: 1000, 
isMoving: true
```

Complete insanity, but if you to find out the hard way, be my guest.

**Comment by christocracy** [comment]
And read this wiki:
https://github.com/transistorsoft/flutter_background_geolocation/wiki/Philosophy-of-Operation

**Comment by gopchi** [comment]
Yes I agree, and in the same time will just cut the part with the timer. Closing for now since we’ll try to speak with the guys and implement it the right way. We’ve got no choice but do it right after all so that it works. Thanks!

**Comment by christocracy** [comment]
It’s never a good idea to use getCurrentPosition on a timer but it shouldn’t crash.  I likely forgot to take care of a threading issue, which is likely solved in the next release.  

Have those who provided the specs take a read of the wiki I posted above.

If you’re aware of how much more strict both iOS and Android are becoming with each new release, it’s almost entirely to prevent people doing the sort of things your current spec requires.

**Comment by fpfilipov94** [comment]
I changed the config of the plugin to match best practices as much as possible. However, it does not work as we expected. We think there is something going on with the motion detection, since the plugin randomly keeps location services off even when the device is definitely moving. Here is the current config:

    geolocationState = await bg.BackgroundGeolocation.ready(bg.Config(
      desiredAccuracy: bg.Config.DESIRED_ACCURACY_HIGH,
      isMoving: true,
      disableElasticity: true,
      distanceFilter: 1,
      activityRecognitionInterval: 5000,
      startOnBoot: true,
      logLevel: bg.Config.LOG_LEVEL_OFF,
      forceReloadOnBoot: true,
      forceReloadOnLocationChange: true,
      stopOnTerminate: false,
      minimumActivityRecognitionConfidence: 55,
      notification: bg.Notification(
        title: 'App Title',
        text: 'Location services active',
      ),
      foregroundService: true,
      persistMode: bg.Config.PERSIST_MODE_NONE,
      maxRecordsToPersist: 0,
    ));

And here is the modified logic to get the location initially and then receive updates:

    // should run only till a fresh location is received
    Timer.periodic(const Duration(seconds: 3), (timer) async {
      if (_busyScanning) return;

      _busyScanning = true;
      try {
        var success = true;

        final l = await bg.BackgroundGeolocation.getCurrentPosition(
          timeout: 2,
        ).catchError((err) {
          success = false;
          print('[INITIAL_LOC_INNER_ERR] $err');
        });

        if (success && l != null) {
          _refreshLocation(l);
          timer.cancel();
        }
      } catch (e) {
        print('[INITIAL_LOC_ERR]');
      }
      _busyScanning = false;
    });

    bg.BackgroundGeolocation.onLocation((bg.Location l) {
      print('ON_LOCATION');
      if (!l.sample) _refreshLocation(l);
    });

We're trying our best to prevent this app from being a battery destroyer, but we still need reliable location updates while the user is moving. Should we perhaps use `activityRecognitionInterval: 0`? What would you suggest?

**Comment by christocracy** [comment]
`activityRecognitionInterval` & `minimumActivityRecognitionConfidence` are no longer used.  they are ignored.

Configure `debug: true, logLevel: bg.Config.LOG_LEVEL_VERBOSE` and go outside for a nice long walk of at least 1km.

**Comment by fpfilipov94** [comment]
Got it. Can you clarify where we can access those logs afterwards?

**Comment by christocracy** [comment]
See Wiki "Debugging"

**Comment by gopchi** [comment]
Here’s something to consider investigating - while the app is in use the location coming from the plugin is alternating. And this is the behaviour with no persisted records - we have disabled that. We are starting to lose hope with using the plugin. It is behaving really strange no matter of the config. 

Also why you even did not answer my initial question from which this issue emerged? When we used getCurrentLocation there was no problem getting the correct location but it crashed on some devices. So I wanted to investigate that and not changing the config and dealing with the battery issues you mentioned.
